0004-bankless-leader.md

```diff
@@ -0,0 +1,88 @@
+---
+simd: '0004'
+title: Bankless Leader
+authors:
+  - Tao Zhu (Solana Labs)
+category: Standard
+type: Core
+status: Draft
+created: 2022-12-08
+feature: (fill in with feature tracking issues once accepted)
+---
+
+## Summary
+
+A bankless leader does the minimum amount of work to produce a valid block. The leader is tasked with ingress transactions, sorting and filtering valid transactions, arranging them into entries, shredding the entries and broadcasting the shreds. While a validator only needs to reassemble the block and replay execution of well formed entries. 
+
+## Motivation
+
+Without having to execute transactions, Bankless leaders can ingest more packets from receiving buffer for filtering and sorting per slort, therefore producing high quality blocks. Some of the benefits include:
+
+1. Process fuller blocks;
+2. Higher prioritized transactions have higher chance to be packed into block earlier;
+3. Combine with Scheduler, to produce more parallizeble entries;
+
+
+## Detailed Design
+
+Normal bank operation for a spend needs to do 2 loads and 2 stores. With this design leader just does 1 load. so 4x less account_db work before generating the block. The store operations are likely to be more expensive than reads.
+
+When replay stage starts processing the same transactions, it can assume that PoH is valid, and that all the entries are safe for parallel execution. The fee accounts that have been loaded to produce the block are likely to still be in memory, so the additional load should be warm and the cost is likely to be amortized.
+
+
+### Fee Account
+
+The [fee account](../terminology.md#fee_account) pays for the transaction to be included in the block. The leader only needs to validate that the fee account has the balance to pay for the fee.
+
+### Balance Cache
+
+For the duration of the leaders consecutive blocks, the leader maintains a temporary balance cache for all the processed fee accounts. The cache is a map of pubkeys to lamports.
+
+At the start of the first block the balance cache is empty. At the end of the last block the cache is destroyed.
+
+The balance cache lookups must reference the same base fork for the entire duration of the cache. At the block boundary, the cache can be reset along with the base fork after replay stage finishes verifying the previous block.
+
+### Balance Check
+
+Prior to the balance check, the leader validates all the signatures in the transaction.
+
+1. Verify the accounts are not in use and BlockHash is valid.
+2. Check if the fee account is present in the cache, or load the account from accounts_db and store the lamport balance in the cache.
+3. If the balance is less than the fee, drop the transaction.
+4. Subtract the fee from the balance.
+5. If fee account is called by system program, or it is passed to a instruction with system program, then reduce its balance to 0 in the cache.
```

Comment:
This should be done for all keys that satisfy 5, so if a system key is spent first then used as a fee payer that the cached balance is zero.